src/wp-includes/class-wp-script-modules.php

```diff
@@ -363,4 +366,119 @@ private function get_src( string $id ): string {
 
 		return $src;
 	}
+
+	/**
+	 * Print data associated with Script Modules.
+	 *
+	 * The data will be embedded in the page HTML and can be read by Script Modules on page load.
+	 *
+	 * @since 6.6.0
+	 *
+	 * Data can be associated with a Script Module via the
+	 * {@see "script_module_data_{$module_id}"} filter.
+	 *
+	 * The data for a Script Module will be serialized as JSON in a script tag with an ID of the
+	 * form `wp-script-module-data-{$module_id}`.
+	 */
+	public function print_script_module_data(): void {
+		$modules = array();
+		foreach ( array_keys( $this->get_marked_for_enqueue() ) as $id ) {
+			$modules[ $id ] = true;
+		}
+		foreach ( array_keys( $this->get_import_map()['imports'] ) as $id ) {
+			$modules[ $id ] = true;
+		}
+
+		foreach ( array_keys( $modules ) as $module_id ) {
+			/**
+			 * Filters data associated with a given Script Module.
+			 *
+			 * Script Modules may require data that is required for initialization or is essential
+			 * to have immediately available on page load. These are suitable use cases for
+			 * this data.
+			 *
+			 * The dynamic portion of the hook name, `$module_id`, refers to the Script Module ID
+			 * that the data is associated with.
+			 *
+			 * This is best suited to pass essential data that must be available to the module for
+			 * initialization or immediately on page load. It does not replace the REST API or
+			 * fetching data from the client.
+			 *
+			 * @example
+			 *   add_filter(
+			 *     'script_module_data_MyScriptModuleID',
+			 *     function ( array $data ): array {
+			 *       $data['script-needs-this-data'] = 'ok';
+			 *       return $data;
+			 *     }
+			 *   );
```

Comment:
My only remaining question is whether adding a filter is the most ergonomic way to add data to a module. Perhaps there should be a function to make this easier? For example:

```php
wp_script_modules()->add_data( 'MyScriptModuleID', array( 'script-needs-this-data' => 'ok' ) );
```

Compare with a typical way that scripts expose data to non-module scripts via `wp_localize_script()`:

```php
wp_localize_script( 'my-script-handle', 'MyScript', array( 'script-needs-this-data' => 'ok' ) );
```

Granted, this use of `wp_localize_script()` is not ideal since it is a hack of what is intended to be used for l10n. What I think is the better more modern alternative is to use `wp_add_inline_script_tag()` (which is also worse in some ways since it requires manual serialization):

```php
wp_add_inline_script(
	'my-script-handle',
	sprintf( 'var MyScript = %s;', wp_json_encode( array( 'script-needs-this-data' => 'ok' ) ) ),
	'before'
);
```

There is also the `WP_Scripts::add_data()` method (and `wp_script_add_data()` helper function) which are used to add "data" to script handles, where `wp_localize_script()` adds this to the `extra` array key, and `wp_add_inline_script()` adds to the `before` and `after` array keys.

What if there was a similar `add_data()` method on `WP_Script_Modules`? 

This could still ultimately get filtered, but it seems maybe a bit strange for the filter to be the primary interface to add data to a module.

Reply:
@gziolo brought this up in the Gutenberg PR: https://github.com/WordPress/gutenberg/pull/61658#issuecomment-2117263472

The motivation for using a filter is that it's a simple interface to implement and it seems sufficient. It doesn't require the class to store any additional data or to handle data merging, the filters handle that. As you note, the filter would be compatible with other methods for adding data if there's demand for them in the future.

The patch you shared is likely sufficient, although it doesn't consider data merging. Should data be merged on subsequent `::add_data()` calls? Maybe there's an argument to merge or replace, or maybe there are separate `::set_data()` and `::add_data()` methods. As the PR stands right now, it avoids having to answer those questions.

I don't feel strongly. If you do, let me know and I can explore adding those methods.

Reply:
I don't feel strongly that they have to be added now, but as the API for the JS-side API is refined to access that data, the PHP-side API could also be refined to set the data.